bee-common/bee-common-derive/src/lib.rs

```diff
@@ -0,0 +1,49 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use proc_macro::{self, TokenStream};
+use proc_macro_error::{abort, proc_macro_error};
+use quote::{quote, ToTokens};
+use syn::{parse_macro_input, Data, DeriveInput, Type};
+
+mod packable;
+
+#[proc_macro_error]
+#[proc_macro_derive(Packable, attributes(packable))]
+pub fn packable(input: TokenStream) -> TokenStream {
+    let DeriveInput {
+        ident,
+        data,
+        attrs,
+        generics,
+        ..
+    } = parse_macro_input!(input);
+    // Check if the type has an `error` attribute.
+    let error_type = packable::parse_attr::<Type>("error", &attrs).map(ToTokens::into_token_stream);
+
+    match data {
+        Data::Struct(data) => {
+            // Use `Infallible` if there was no error attribute.
+            let error_type = error_type.unwrap_or_else(|| quote!(core::convert::Infallible));
+            // Generate the implementation for the struct.
+            let (pack, packed_len, unpack) = packable::gen_bodies_for_struct(data.fields);
+            packable::gen_impl(&ident, &generics, error_type, pack, unpack, packed_len).into()
+        }
+        Data::Enum(data) => {
+            // Verify that the enum has a `"tag_ty"` attribute for the type of the tag.
+            let tag_ty = packable::parse_attr::<Type>("tag_ty", &attrs).unwrap_or_else(|| {
```

Comment:
I'm not a big fan of `tag_ty` because I like expressive attributes, shortening make it harder to read IMO, as a comparison serde doesn't shorten anything, I'd prefer the full `tag_type` tbh. Another thing is that `type` is a keyword so we took the habit of calling these tag bytes `kind`, not sure if we want to also call it that here, we should discuss it.

Reply:
I'll use `tag_type` for now. The thing with kind is that you don't know what you should put there.